lib/models/NotificationConfiguration.js

```diff
@@ -0,0 +1,245 @@
+            if (!supportedNotificationEvents.includes(e)) {
+                eventsObj.error = errors.MalformedXML.customizeDescription(
+                    'event array contains invalid or unsupported event');
+            }
+            eventsObj.events.push(e);
+        });
+        return eventsObj;
+    }
+
+    /**
+     * Check that filter array is valid
+     * @param {array} filter - filter array
+     * @return {object} - contains error if parsing failed or filter array
+     */
+    _parseFilter(filter) {
+        const filterObj = {
+            filterRules: [],
+        };
+        if (filter && filter[0]) {
+            if (!filter[0].S3Key || !filter[0].S3Key[0]) {
+                filterObj.error = errors.MalformedXML.customizeDescription(
+                    'if included, queue configuration filter must contain S3Key');
+                return filterObj;
+            }
+            const filterRules = filter[0].S3Key[0];
+            if (!filterRules.FilterRule || !filterRules.FilterRule[0]) {
+                filterObj.error = errors.MalformedXML.customizeDescription(
+                    'if included, queue configuration filter must contain a rule');
+                return filterObj;
+            }
+            const ruleArray = filterRules.FilterRule;
+            ruleArray.forEach(r => {
+                if (!r.Name || !r.Name[0] || !r.Value || !r.Value[0]) {
+                    filterObj.error = errors.MalformedXML.customizeDescription(
+                        'each included filter must contain a name and value');
+                } else if (!['Prefix', 'Suffix'].includes(r.Name[0])) {
+                    filterObj.error = errors.MalformedXML.customizeDescription(
+                        'filter Name must be one of Prefix or Suffix');
+                } else {
+                    filterObj.filterRules.push({
+                        name: r.Name[0],
+                        value: r.Value[0],
+                    });
+                }
+            });
+        }
+        return filterObj;
+    }
+
+    /**
+     * Check that id string is valid
+     * @param {string} id - id string (optional)
+     * @return {object} - contains error if parsing failed or id
+     */
+    _parseId(id) {
+        const idObj = {};
+        idObj.propName = 'ruleID';
+        if (id && id[0].length > 255) {
+            idObj.error = errors.InvalidArgument.customizeDescription(
+                'queue configuration ID is greater than 255 characters long');
+            return idObj;
+        }
+        if (!id || !id[0] || id[0] === '') {
+            // id is optional property, so create one if not provided or is ''
+            // We generate 48-character alphanumeric, unique id for rule
+            idObj.id = Buffer.from(UUID.v4()).toString('base64');
+        } else {
+            idObj.id = id[0];
+        }
+        // Each ID in a list of rules must be unique.
+        if (this._ids.includes(idObj.id)) {
```

Comment:
If `this._ids` is only used to check for uniqueness I would use a Set

Reply:
ooh something I forgot to say is that I think `_ids` could be a little more specific, maybe like `_existingIds`. I thought the side effect of adding to `_ids` was a bigger deal until I realized its just used for this.